src/libstd/collections/hash/raw/mod.rs

```diff
@@ -0,0 +1,1210 @@
+            Self::new()
+        } else {
+            unsafe {
+                let mut new_table = ManuallyDrop::new(
+                    Self::new_uninitialized(self.buckets(), Fallibility::Infallible)
+                        .unwrap_or_else(|_| hint::unreachable_unchecked()),
+                );
+
+                // Copy the control bytes unchanged. We do this in a single pass
+                self.ctrl(0)
+                    .copy_to_nonoverlapping(new_table.ctrl(0), self.buckets() + Group::WIDTH);
+
+                {
+                    // The cloning of elements may panic, in which case we need
+                    // to make sure we drop only the elements that have been
+                    // cloned so far.
+                    let mut guard = guard((0, &mut new_table), |(index, new_table)| {
+                        if mem::needs_drop::<T>() {
+                            for i in 0..=*index {
+                                if is_full(*new_table.ctrl(i)) {
+                                    new_table.bucket(i).drop();
+                                }
+                            }
+                        }
+                        new_table.free_buckets();
+                    });
+
+                    for from in self.iter() {
+                        let index = self.bucket_index(&from);
+                        let to = guard.1.bucket(index);
+                        to.write(from.as_ref().clone());
+
+                        // Update the index in case we need to unwind.
+                        guard.0 = index;
+                    }
+
+                    // Successfully cloned all items, no need to clean up.
+                    mem::forget(guard);
+                }
+
+                // Return the newly created table.
+                new_table.items = self.items;
+                new_table.growth_left = self.growth_left;
+                ManuallyDrop::into_inner(new_table)
+            }
+        }
+    }
+}
+
+unsafe impl<#[may_dangle] T> Drop for RawTable<T> {
+    #[inline]
+    fn drop(&mut self) {
+        if self.bucket_mask != 0 {
+            unsafe {
+                if mem::needs_drop::<T>() {
+                    for item in self.iter() {
+                        item.drop();
+                    }
+                }
+                self.free_buckets();
+            }
+        }
+    }
+}
+
+impl<T> IntoIterator for RawTable<T> {
+    type Item = T;
+    type IntoIter = RawIntoIter<T>;
+
+    #[inline]
+    fn into_iter(self) -> RawIntoIter<T> {
+        unsafe {
+            let iter = self.iter();
+            let alloc = self.into_alloc();
+            RawIntoIter { iter, alloc }
+        }
+    }
+}
+
+/// Iterator over a a sub-range of a table. Unlike `RawIter` this iterator does
+/// not track an item count.
+pub struct RawIterRange<T> {
+    // Using *const here for covariance
+    data: *const T,
+    ctrl: *const u8,
+    current_group: BitMask,
+    end: *const u8,
```

Comment:
`RawIter` stores the number of items so it does not need to know the end of the metadata array. `RawIterRange` needs to be able to split into subparts of the hashtables and that requires `end`. I think this splitting is only needed by rayon.

Also note that in the presence of a lot of deletions an iterator with item count can be potentially faster: it can exist as soon as item count goes to zero - which might be halfway through the metadata array.

Reply:
I split the item count out as a separate field so that it can easily be eliminated by the compiler if it is not used. In general iterators will be completely inlined and optimized by the compiler.

Reply:
Can we make `RawIter` use item count and not use `end` so that `end` is eliminated? The common case is to take a `RawIter` and iterate it to the end and using item count to terminate the iterate is the optimal strategy (using `end` is slower).

Reply:
Are you sure that using `end` is actually slower? In 99% of cases the entries will be evenly distributed across the table, and using the item counter doesn't help.

The reason I chose an end pointer over an item count is that the item count needs to be updated on each element (inner loop) whereas the end pointer only needs to be checked when moving to another group (outer loop).

Reply:
It will be much better when a table is sparse (full of deletions). When a table doesn't have deletions both options should be equivalent. Admittedly sparse tables are very rare.

Note: if you resurrect the SENTINEL meta then you can terminate on that instead of `end` and you will not need the `end` in the case where you iterate over the whole table. But that won't work for rayon.

Reply:
Can you elaborate on how this makes matching logic slower?

Reply:
Currently `match_empty_or_deleted` only needs a single instruction since it just looks at the top bit of each byte. However if it has to exclude SENTINEL values then 3 instructions are required.

Reply:
I see. Do you remember off hand how much difference does it make?